spi-proposal.md

```diff
@@ -0,0 +1,328 @@
+3 = 3rd part of sequence, etc...
+... up to 100, after 100, wrap back to 1
+
+127 = end sequence
+```
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_TRANSFER_REQUEST (0x01)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel)
+4:  sequence             (see note on sequence above)
+5:  numBytes to read and write
+6:  data 0               (bits 0-6)
+7:  data 1               (bit 7) // or inverted depending on bitOrder setting
+... up to numBytes * (8 / 7)
+N:  END_SYSEX
+```
+
+### SPI Read only
+
+Use if device has a ready signal or requires a delay after a write command.
+Writes `0x00` for each byte to be read. Respond with a `SPI_REPLY` message.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_READ_REQUEST     (0x02)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel)
+4:  sequence             (see note on sequence in Transfer section)
+5:  numBytes to read
+6:  END_SYSEX
+```
+
+### SPI Write only
+
+Use when a transfer does not require a response.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_WRITE            (0x03)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel)
+4:  sequence             (see note on sequence in Transfer section)
+5:  numBytes to write
+6:  data 0               (bits 0-6)
+7:  data 1               (bit 7) // or inverted depending on bitOrder setting
+... up to numBytes * (8 / 7)
+N: END_SYSEX
+```
+
+### Reply
+
+A byte array of data received from the SPI slave device in response to the
+transfer. Used for both `SPI_TRANSFER_REQUEST` AND `SPI_READ_REQUEST`.
+
+**sequence**
+
+The sequence byte specifies which part of a sequence the reply corresponds to.
+Helps ensure reply matches request.
+
+```
+0 = single transfer (no sequence)
+
+1 = start sequence (1st part of sequence)
+2 = 2nd part of sequence
+3 = 3rd part of sequence, etc...
+... up to 100, after 100, wrap back to 1
+
+127 = end sequence
+```
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_REPLY            (0x04)
+3:  deviceId | channel   (bits 2-6: deviceId, bits 0-1: channel)
+4:  sequence             (see note on sequence above)
+5:  numBytes in reply
+6:  data 0               (bits 0-6)
+7:  data 1               (bit 7) // or inverted depending on bitOrder setting
+... up to numBytes * (8 / 7)
+N: END_SYSEX
+```
+
+### Begin
+
+Call once to initialize SPI hardware. Can optionally be set internally upon
+receiving first `SPI_CONFIG_DEVICE` message.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_BEGIN            (0x05)
+3:  channel              (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+### End
+
+Call once to release SPI hardware send before quitting a Firmata client
+application.
+
+```
+0:  START_SYSEX
+1:  SPI_DATA             (0x68)
+2:  SPI_END              (0x06)
+3:  channel              (HW supports multiple SPI ports. range = 0-3, default = 0)
+4:  END_SYSEX
+```
+
+# Addendum
```

Comment:
Not committed to the commands described in this section yet. Note the two different options. Curious what others think. This is also something that could be added later if necessary, or on the other hand it could be a very simple way to structure SPI for Firmata, but it does require sending a lot more messages.